### PR TITLE
Move our base controller infrastructure into knative/pkg.

### PR DESCRIPTION
**Pull request by mattmoor**
The `controller.go` is from: https://github.com/knative/serving/pull/1770, however, this adds 100% coverage of `controller.go`, which we have been missing in `knative/serving`.

Fixes: https://github.com/knative/pkg/issues/8

WIP until https://github.com/knative/serving/pull/1770 lands.

### CODE REVIEW

**Comment by mattmoor** [comment]
cc @vaikas-google @n3wscott @evankanderson @grantr

**Comment by mattmoor** [comment]
cc @dprotaso

**Comment by mattmoor** [comment]
cc @ImJasonH

File: controller/controller.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+	// time, and makes it easy to ensure we are never processing the same item
+	// simultaneously in two different workers.
+	WorkQueue workqueue.RateLimitingInterface
+
+	// Sugared logger is easier to use but is not as performant as the
+	// raw logger. In performance critical paths, call logger.Desugar()
+	// and use the returned raw logger instead. In addition to the
+	// performance benefits, raw logger also preserves type-safety at
+	// the expense of slightly greater verbosity.
+	logger *zap.SugaredLogger
```
Comment (n3wscott) [inline controller/controller.go]: should this be a logger or a context?

**Comment by n3wscott** [comment]
this could land before knative/serving#1770  because the dep is inverted, right?

File: controller/controller.go (hunk 1 above)
Comment (mattmoor) [reply controller/controller.go]: I don't think so?  I am assuming this is in reference to my comment on the other PR?

**Comment by n3wscott** [comment]
goooo innn PR...

**Pull request merged**